server/jamovi/tests/test_dataset.py

```diff
@@ -1,60 +1,214 @@
-    column = ds.append_column("fred")
-    column.set_data_type(DataType.TEXT)
+    # GIVEN a populated column
+    column = add_column_to_dataset(ds, measure_type, values)
 
-    # WHEN setting values
-    column.set_value(0, "fred")
-    column.set_value(1, "44")
-    column.set_value(2, "jim")
-    column.set_value(3, "33.1")
+    for index, value in enumerate(values):
+        assert equals(column.get_value(index), value)
 
-    # THEN these are persisted
-    assert column.get_value(0) == "fred"
-    assert column.get_value(1) == "44"
-    assert column.get_value(2) == "jim"
-    assert column.get_value(3) == "33.1"
+    levels_must_equal(column.dlevels, expected_levels)
 
 
+@pytest.mark.parametrize(
+    ("measure_type", "values", "edits", "expected_levels"),
+    [
+        (
+            MeasureType.NOMINAL,
+            ["fred", "44", "jim", "33.1", "", "33.1", "fred"],
+            [NAN_INT, NAN_INT, 2, None, None, None, None],
+            [
+                {
+                    "value": 0,
+                    "label": "fred",
+                    "import_value": "fred",
+                    "pinned": False,
+                    "count": 1,
+                    "count_ex_filtered": 1,
+                },
+                {
+                    "value": 1,
+                    "label": "33.1",
+                    "import_value": "33.1",
+                    "pinned": False,
+                    "count": 3,
+                    "count_ex_filtered": 3,
+                },
+            ],
+        ),
+        (
+            MeasureType.NOMINAL,
+            [2000, 500, 500, 500, 1000, 1000, 2000, NAN_INT, 500],
+            [None, 500, 2000, NAN_INT, 2000, NAN_INT, NAN_INT, 500, 500],
+            [
+                {
+                    "value": 500,
+                    "label": "500",
+                    "import_value": "500",
+                    "pinned": False,
+                    "count": 3,
+                    "count_ex_filtered": 3,
+                },
+                {
+                    "value": 2000,
+                    "label": "2000",
+                    "import_value": "2000",
+                    "pinned": False,
+                    "count": 3,
+                    "count_ex_filtered": 3,
+                },
+            ],
+        ),
+    ],
+)
+def test_levels_are_trimmed_when_counts_reach_zero(
+    empty_dataset: DuckDataSet,
+    measure_type: MeasureType,
+    values: Sequence[int | str],
+    edits: Sequence[int | str | None],
+    expected_levels: Sequence[dict[str, Any]],
+):
+    """test that levels are trimmed accordingly"""
+    ds = empty_dataset
+    ds.set_row_count(len(values))
+
+    # GIVEN a populated column
+    column = add_column_to_dataset(ds, measure_type, values)
+
+    for index, value in enumerate(edits):
+        if value is not None:
+            column.set_value(index, value)
+
+    levels_must_equal(column.dlevels, expected_levels)
+
+
+@pytest.mark.skip(reason="TODO")
```

Comment:
Is this still a TODO?

Reply:
yeah ... changing column types is the last thing on my TODO list